### ISSUE TITLE
right way to "include()" in $http->on("request", ...)

### ISSUE THREAD

**Issue by flddr** [swoole/swoole-src#1963]
Whats the right way to **include(_filename_);** from _ramfs/disk_ in the following code?

```php
<?php

$http = new swoole_http_server("127.0.0.1", 9501);

$http->on("request", function ($request, $response) {

  # ho to use include(/media/ramfs/filename) ?

});

$http->start();
?>
```

I need **include(_filename_);** (_because it parses the file as PHP and so on_) - and i need _ramfs_, too.

Edit: i mean because of _async_, so only way is coroutine?

**Comment by twose** [comment]
normally you just need to know that you should add `require "autoload.php"` in `workerStart` event.

every file would require only once, and it will keep in memory.

**Comment by flddr** [comment]
@twoose,

i need to use **include** as HTML-template-parser in _every_ request.

There is **_no need_** for a complete overwhelming _template-engine_, because **require/include** processes HTML/PHP perfect, including all scope variables etc.

So i just need the correct _async_ form to get those files from my disk / ramdisk :) I do not want to _once.

Must it be done via coroutine? :)

**Comment by flddr** [comment]
Personally i did not found a good way to _eval_ such file-strings. _eval_ needs extra tagging and is error-prone, whereas _included_ files can be normal HTML/PHP files; you can simply edit in every HTML-editor and there is no problem in PHP-parsing - you can do whatever you want - whereas template-engines are almost laggy in feautures

**Comment by twose** [comment]
if you want to move codes on php-fpm to swoole? it's too difficult, no global vars, no static vars.
if you just want to use the template, just `ob_start` and `ob_get_clean`.

**Comment by flddr** [comment]
no, i just want to include a file, like index.html for GET /

But, like file_get_contents, include is blocking?

**Comment by flddr** [comment]
file_get_contents or include works fine, but swoole wrote somewhere not to use it? Or with coroutine? I did not know actually

**Comment by flddr** [comment]
Just the question how to use **include** or **file_get_contents** correct in async. Can i use them out of the box in `$http->on("request", ...` ?

:)

**Comment by twose** [comment]
it(include file) doesn't matter If you don't abuse it.

**Comment by flddr** [comment]
Hmm, i will make some tests and try over time :) thank you

**Comment by flddr** [comment]
I have done some tests, and benchmark for all is nearly same requests/s:

1) `$teststring = "hello world"`
2) `file_get_contents()` `# "hello world"`
3) `include()` `# "hello world"`

from _ramdisk and ssd_. With 10 bytes or 30k bytes and concurrency 100.

So i guess callbacks like `$http->on("request", ...` are each already done in single **coroutines**? @twose
